ui/src/wallet.ts

```diff
@@ -91,6 +72,40 @@ export function build(
     status: Status.NotConnected,
   });
 
+  function onModalHide(): void {
+    const store = svelteStore.get(stateStore);
+    if (store.status === Status.NotConnected) {
+      reinitWalletConnect();
+    }
+  }
+
+  function newWalletConnect(): WalletConnect {
+    return new WalletConnect({
+      bridge: "https://bridge.walletconnect.org",
+      qrcodeModal: qrCodeModal,
+    });
+  }
+
+  // We need to reinitialize `WalletConnect` until this issue is fixed:
+  // https://github.com/WalletConnect/walletconnect-monorepo/pull/370
+  function reinitWalletConnect() {
+    walletConnect = newWalletConnect();
+    signer.walletConnect = walletConnect;
+  }
+
+  const qrCodeModal = {
+    open: (uri: string, _cb: unknown, _opts?: unknown) => {
+      uriStore.set(uri);
+      modal.toggle(path.walletQRCode(), onModalHide);
+    },
+    close: () => {
+      // N.B: this is actually called when the connection is established,
+      // not when the modal is closed per se.
+      stateStore.set({ status: Status.Connecting });
```

Comment:
Setting the status to “connecting” seems to contradict the comment above. Shouldn’t it be `Status.Connected`? And why wasn’t this there before?

Reply:
> Setting the status to “connecting” seems to contradict the comment above

Indeed. I think rephrasing the comment to the following is more correct:

> // N.B: this is actually called when the connection is initiated,
> // not when the modal is closed per se.

Meaning, when the user scans the QR Code and the connection gets going, this closure is called.

> Shouldn’t it be `Status.Connected`

Ideally, this closure would be `onConnected` or something alike. But the actual connection establishment happens in `walletConnect.connect` (line 125). When it succeeds, we then call `await initialize();`, line 135.

> And why wasn’t this there before?

Because until the need for this workaround, we didn't need to eagerly set the status to what it factually is but to which is not yet set at this point. i.e, this closure is called before the `walletConnect.connect()` promise resolves. In that time in between, `modal.hide()` is called, which now calls the passed `onHide` parameter. When that happens, we need to know whether the modal was closed as a dismissal (the user just closed it) or if it was closed because the connection is being established. We can only capture the latter, in this very closure. So, setting the status to `Connecting` eagerly here allows us to check it in `onModalHide`; if it's `Connecting`, we don't do anything. If it's `NOT_CONNECTED`, then we know the user has dismissed the modal and that we must, therefore, reinitialize `walletConnect`.